sdk/azidentity/chained_token_credential.go

```diff
@@ -70,15 +67,12 @@ func NewChainedTokenCredential(sources []azcore.TokenCredential, options *Chaine
 func (c *ChainedTokenCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (*azcore.AccessToken, error) {
 	if !c.retrySources {
 		// ensure only one goroutine at a time iterates the sources and perhaps sets c.successfulCredential
+		c.cond.L.Lock()
 		for {
-			c.mut.RLock()
-			mustIterate := c.successfulCredential == nil
-			c.mut.RUnlock()
-			if !mustIterate {
+			if c.successfulCredential != nil {
+				c.cond.L.Unlock()
 				return c.successfulCredential.GetToken(ctx, opts)
 			}
-			c.cond.L.Lock()
-			defer c.cond.L.Unlock()
 			if !c.iterating {
 				c.iterating = true
```

Comment:
I think you need to `c.cond.L.Unlock()` after this assignment so the remaining goroutines unblock from line 70 and block on `c.cond.Wait()`.

Reply:
That would more efficiently get goroutines to the `Wait()` but it reintroduces the race because the iterating goroutine would assign the variable while not holding the lock.

Reply:
But all the other goroutines would be waiting yes?

Reply:
Hmm I suppose not, if one arrives late.

Reply:
I'm with you in believing it ought to be safe to unlock here because only the goroutine which does the assignment would iterate, while all others wait for the signal. However, unlocking here triggers the race detector, so either it or me isn't smart enough to figure this out 😆

Reply:
Ah yes, I get it now.  The lock must be acquired before writing to shared state (just like we do [here](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/internal/shared/expiring_resource.go#L95)).  I think that after the call to `cred.GetToken()` you need to do this.
```go
if c.iterating {
    c.cond.L.Lock()
}
```
What do you think?

Reply:
Yes, that does the trick. A goroutine can iterate without holding the lock but must acquire the lock afterward.